scivision/catalog/catalog.py

```diff
@@ -1,85 +1,221 @@
-        """Query the Pandas dataframe."""
-        queries = [f"{k} == '{v}'" for k, v in query.items()]
-        query_str = " & ".join(queries)
-        result = self._database.query(query_str)
-        return result.to_dict("records")
+    def compatible_models(self, datasource) -> pd.DataFrame:
+        """
+        Return all models that are compatible with datasource
 
-    @property
-    def keys(self) -> List[str]:
-        """Return the query keys."""
-        return self._database.columns.tolist()
+        Parameters
+        ----------
+        datasource : str or dict-like
 
-    def values(self, key: str) -> List[str]:
-        """Return the unique values for a query key."""
-        if key not in self.keys:
-            raise ValueError(f"Key {key} not found.")
-        values = self._database[key].tolist()
-        return list(set(values))
+        Any dictionary-like (including CatalogDatasourceEntry)
+        that has keys 'format', 'tasks' and 'labels', representing these properties of the datasource.
 
+        If a string is passed, this is used to look up the datasource (in `self._datasources`).
 
-_catalog = PandasCatalog()
+        Returns
+        -------
+        result: pd.Dataframe
 
+        A dataframe containing the models compatible with the given datasource
+        """
+        if isinstance(datasource, str):
+            return self._compatible_models(
+                self._datasources.set_index("name").loc[datasource]
+            )
+        else:
+            return self._compatible_models(datasource)
 
-@koala
-def query(query: Dict[str, str]) -> list:
-    """Search the catalog using the query.
+    def compatible_datasources(self, model) -> pd.DataFrame:
+        """
+        Return all models that are compatible with datasource
 
-    Parameters
-    ----------
-    query : dict
-        A dictionary describing the search query as key value pairs.
+        Parameters
+        ----------
+        model : str or dict-like
 
-    Returns
-    -------
-    result : list
-        A list of catalog entries matching the search requirements.
-    """
+        Any dictionary-like (including CatalogModelEntry)
+        that has keys 'format', 'tasks' and 'labels_required', representing these properties of the model.
 
-    result = _catalog.query(query)
-    return result
+        If a string is passed, this is used to look up the model (in `self._models`).
 
+        Returns
+        -------
+        result: pd.Dataframe
 
-@koala
-def keys() -> List[str]:
-    return _catalog.keys
+        A dataframe containing the datasources compatible with the given model
+        """
+        if isinstance(model, str):
+            return self._compatible_datasources(
+                self._models.set_index("name").loc[model]
+            )
+        else:
+            return self._compatible_datasources(model)
 
 
-@koala
-def values(key: str) -> List[str]:
-    return _catalog.values(key)
+default_catalog = PandasCatalog()
```

Comment:
I think this is neat :+1:

If I were to work on the `add_model` and `add_dataset` functions again after these changes, do you envision people doing something like:

1. `default_catalog.add_model('my_model')`
2. `default_catalog.compatible_datasources('my_model')`

If so, maybe we'd rather just call it `catalog` instead of `default_catalog`? (I'm not sure)

Reply:
Yes exactly.

For that particular usage, you can also do something like:
```
default_catalog.compatible_datasources(CatalogModelEntry(...))
```
without modifying the default catalog.

Was also not completely sure what to call it.  Maybe catalog is better, but there might be some confusion with the module named `catalog`.

Reply:
(writing that out makes `CatalogModelEntry` seem a bit clunky - could be worth thinking of a better name)

Reply:
In that case, perhaps if `compatible_datasources` detects that you're providing a dictionary of a model that isn't in the catalog, it should just turn the dict into a `CatalogModelEntry`, allowing you return a result?

Reply:
e.g. there's an `if type(model) == Dict` statement that creates `CatalogModelEntry`

Reply:
It actually does something close to that: it accepts any argument with __getitem__ and the relevant keys, so a dictionary would also work (no need for the type/isinstance check).

I'm not completely satisfied with this, though - if you can pass a dictionary, then why not just allow keyword arguments? I couldn't get this quite right either, so decided to leave it for now (probably part of the same future PR as the query replacement)